### PR TITLE
Set IsRequired on ApiDescriptions for endpoints

### LINKED ISSUE #35081: Set required parameter attribute in OpenAPI metadata
By default, OpenAPI assumes that all parameters are optional parameters.

https://swagger.io/docs/specification/describing-parameters/

We don't currently propagate any information that says `required: true` for parameters that are required in an endpoint

This builds on top of the work done in #32375

### PR DESCRIPTION
**Pull request by davidfowl**
- Use the same logic we have in RequestDelegateFactory.Create to
determine if a method parameter is required or not. We then set the
IsRequired property on the ApiParameterDesciption.
- Added tests

Fixes https://github.com/dotnet/aspnetcore/issues/35081

~TODO: Figure out why tests don't properly expose NRT metadata. It might be because of this bug https://github.com/dotnet/roslyn/issues/55254~

### CODE REVIEW

File: src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs [hunk 1]
```diff
@@ -324,18 +327,65 @@ public void AddsMultipleParameters()
...
+            Assert.Equal(typeof(InferredJsonClass), fromBodyParam0.ModelMetadata.ModelType);
+            Assert.Equal(BindingSource.Body, fromBodyParam0.Source);
+            Assert.False(fromBodyParam0.IsRequired);
+
+            var fromBodyParam1 = apiDescription1.ParameterDescriptions[0];
+            Assert.Equal(typeof(InferredJsonClass), fromBodyParam1.Type);
+            Assert.Equal(typeof(InferredJsonClass), fromBodyParam1.ModelMetadata.ModelType);
+            Assert.Equal(BindingSource.Body, fromBodyParam1.Source);
+            Assert.False(fromBodyParam1.IsRequired);
+        }
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:385]: I looked at this because I know we have a similar `RequestDelegateFactoryTest` that passes. Specifically [RequestDelegateHandlesBodyParamOptionality](https://github.com/dotnet/aspnetcore/blob/fa9dd9c5fe476706e33819ded8b96bf72a5fef41/src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs#L1565) At first, I thought it was just because that test used a local function rather than a lambda, so I was going to suggest switching to that for now until https://github.com/dotnet/roslyn/issues/55254 is fixed. I turns out switching to a local function didn't fix the test though.

Guess what *does* fix this test!?

```suggestion
        }
                
        [Theory]
        [InlineData(null)]
        public void RandomTheory(object ignore)
        {
        }
```

I don't know entirely why. It doesn't really change anything about the IL generated for the `TestIsRequiredFromBody` test this somehow fixes.

What I did notice is adding this same `RandomTheory` method outside of the `#nullable enable` section *does not* fix this even though the theory doesn't mark anything as nullable.

Moving the `RandomTheory` into the nullable section does change the IL for `RandomTheory` by adding the `NullableContextAttribute` to the method, but I don't know why this affects the `TestIsRequiredFromBody` which doesn't have the attribute before or after the fix.

image

Comment (davidfowl) [reply src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:385]: It's this bloody compiler bug

Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:385]: Here's a simpler "fix".

```suggestion
        }
        
        private object RandomMethod() => new object();
```

It appears having any method in the nullable section with either a parameter or non-void return type does it as that's required to get the `NullableContextAttribute` somewhere and that apparently is enough to get nullability to work even for the other method without the attribute.

**Review by halter73 — APPROVED** [review]

File: src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs (hunk 1 above)
Comment (JamesNK) [reply src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:385]: Is this at all related to https://github.com/dotnet/aspnetcore/issues/32804?

tldr; there is an option configured that only outputs nullable metadata on public types - https://github.com/dotnet/aspnetcore/issues/32804#issuecomment-842791651

Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:385]: I don't think so. This PR removed `nullablePublicOnly` from the `Features` property in the test project. https://github.com/dotnet/aspnetcore/pull/35233/files#diff-b73478a77d203a214533c8257af02a208e1d529c629f682f2e2c2999cac0da05R6
